File: CodeChallenges/bs-sqrt.ts
function mySqrt (x: number): number {
    let l = 0;
    let r = x;
    while (l <= r) {
        let mid = Math.floor((l + r) / 2);
        let squre = mid * mid;
        if (squre > x) {
            r = mid - 1;
        } else if (squre < x) {
            l = mid + 1;
        } else {
            return mid;
        }
    }
    return r;
}

console.log(mySqrt(12));

File: CodeChallenges/bs-sqrt.java
package CodeChallenges;

class Sqrt {
    public int mySqrt(int x) {
        long l = 0;
        long r = x;
        while (l <= r) {
            long mid = (l + r) / 2;
            long guess = mid * mid;
            if (guess > x) {
                r = mid - 1;
            } else if (guess < x) {
                l = mid + 1;
            } else {
                return (int) mid;
            }
        }
        return (int) r;
    }
}
